### ISSUE TITLE
unexpected "the substitution forms a cycle"

### ISSUE THREAD

**Issue by gelisam** [gelisam/typelevel-rewrite-rules#15]
when typechecking this code:

```haskell
ex2g :: (as ++ (bs ++ cs)) ~ cs
     => (((as ++ bs) ++ cs) ~ cs => r)
     -> proxy as
     -> proxy bs
     -> proxy cs
     -> r
ex2g r _ _ _ = r
```

**Comment by tscholak** [comment]
oh, I was just going to report that. I get the same unexpected error when I define a right associative law.

**Comment by gelisam** [comment]
We figured out why the above example causes this error message.

GHC sometimes splits type family applications into separate equations, so

```
[W] ((as ++ bs) ++ cs) ~ cs
```

might become

```haskell
[G] (as ++ bs) ~ fsk0
[W] (fsk0 ++ cs) ~ cs
```

This is problematic for this plugin, since we want to rewrite `(as ++ bs) ++ cs` to `as ++ (bs ++ cs)` but `fsk0 ++ cs` doesn't match the rewrite rule. So we collect all the `something ~ var` givens into `typeSubst`, and splice their definitions back in before applying the rewrite rules:

```haskell
typeSubst = [([| fsk0 |], [| as ++ bs |])]
applyTypeSubst typeSubst [|fsk0 ++ cs|] = [| (as ++ bs) ++ cs |]
applyRules [| (as ++ bs) ++ cs |] = [| as ++ (bs ++ cs) |]
```

In the problematic example, however, `applyTypeSubst` never terminates because of this given from the type signature:

```haskell
[G] (as ++ (bs ++ cs)) ~ cs
```

This `cs` is erroneously assumed to be one of those `fsk0` type variables, and thus ends up in the `typeSubst`. Then `applyTypeSubst` whenever it is applied to an expression containing `cs`, because after replacing `cs` with `as ++ (bs ++ cs)`, the result still has an occurrence of `cs`, which is expanded again, and so on.

```haskell
cs
as ++ (bs ++ cs)
as ++ (bs ++ (as ++ (bs ++ cs)))
as ++ (bs ++ (as ++ (bs ++ (as ++ (bs ++ cs))))
...
```

Fortunately, `applyTypeSubst` detects this loop and fails with "the substitution forms a cycle".

**Comment by gelisam** [comment]
One way forward would be to somehow distinguish between the givens generated by GHC and those written by the user. However, there is one test case which relies on the existing behaviour:

```haskell
ex2f :: forall proxy xs as bs cs r. xs ~ ((as ++ bs) ++ cs)
     => (xs ~ (as ++ (bs ++ cs)) => r)
     -> proxy as
     -> proxy bs
     -> proxy cs
     -> r
ex2f r _ _ _ = r
```

This time the wanted `xs ~ (as ++ (bs ++ cs))` is already in normal form, while the given `xs ~ ((as ++ bs) ++ cs)` is not. Since we do not yet simplify the givens, this seems like this should fail, but thanks to `applyTypeSubst`, the wanted gets expanded to `((as ++ bs) ++ cs) ~ (as ++ (bs ++ cs))`, then gets rewritten to `(as ++ (bs ++ cs)) ~ (as ++ (bs ++ cs))`, and then gets solved.

So let's revisit this after we simplify the givens in #7.
